### ISSUE TITLE
import is very slow

### ISSUE THREAD

**Issue by glyph** [brython-dev/brython#222]
I tried to report this on #123 along with a whole bunch of other related issues, but I think it would be good to have a separate issue that calls out this one specific thing.

Importing modules on brython is slow.  Unacceptably slow, in my opinion, because one of the major advantages of Python over JS for in-browser development is the very nice module system, and brython punishes you for using it with very slow import times.

The solution proposed on #123 is to implement a way to pre-generate `.pyc.js` files, since the transpiler is a clear bottleneck in import; but that is a proposed solution, so this issue is specifically for the *problem* of import being slow, if others wish to address it some other way.

**Comment by olemis** [comment]
The fact that importing py modules should be slow is definitely "obvious" considering that firstly module code lookup (e.g. sys.path) and download is needed, followed by py to js translation and finally running js code . I think you are talking of getting rid of the two first steps , isn't it ?

**Comment by olemis** [comment]
After adding support for PEP-302 hooks in #189 it seems it should not be quite difficult to get this done. @glyph could you please prepare a simple package structure consisting of compiled js modules ? I might need it for testing purposes . Thanks in advance !

**Comment by glyph** [comment]
@olemis - I'm not sure what you mean by "prepare a simple package structure" - I don't know how to run the translator in any kind of ahead-of-time mode, indeed as I understand it isn't currently possible, and I don't know enough about the code to implement it.  What exactly would you like me to prepare?

**Comment by olemis** [comment]
"Running the translator" is relatively easy 

  * Go to console in www/tests/console.html
  * Edit it by changing `onload=brython()` to `onload=brython(10)`
  * After executing `import module` statements in Brython console , generated js module code will be output in the browser console (e.g. Console tab in Chrome Dev Tools)

I ask you this favor because I'm working on some tickets in parallel . If you help me with this I'll be able to finish other tickets and move faster ;) . Thanks in advance !

**Comment by glyph** [comment]
@olemis - no promises, I'm quite busy with other things, but I'll try to figure this out :).

**Comment by glyph** [comment]
@olemis - that commit looks awesome - does it work? :)

**Comment by olemis** [comment]
Note quite . There are still [pending issues](https://groups.google.com/d/topic/brython/1qzxCvaSfDY‎)

**Comment by PierreQuentel** [comment]
I have been working on py_import and py_import_hooks these days, I will commit the result to the master branch probably today

**Comment by glyph** [comment]
I'm excited to see this work! Let me know what I can do to test.

@olemis - I tried following that link but google groups tells me "the input is invalid".

**Comment by PierreQuentel** [comment]
I committed a set of changes today, including those for the import mechanism. The main change is that finders are implemented as classes, with classmethods, as in standard Python. I also changed the list of finders in sys.meta_path depending on some options (static_stdlib_import, use of the Virtual File Systsme) to limit the number of Ajax calls.

Olemis, I'm not sure it fixes all the bugs you found (for me too the link doesn't work), can you try ? It fixes at least all those I had found.

**Comment by olemis** [comment]
I'll pull those changes tomorrow .

@glyph @PierreQuentel can you see [this link](https://groups.google.com/d/msg/brython/1qzxCvaSfDY/5F4hzUWd1LkJ) ?

**Comment by PierreQuentel** [comment]
Yes, it leads to a thread in the Google Group

**Comment by olemis** [comment]
@glyph after the changes committed by @PierreQuentel it's definitely looking better , yet untested . I need a python module and it's corresponding compiled form to write a test and be sure . Immediately after I'll merge my branch with trunk .

**Comment by glyph** [comment]
:clap: :clap: :clap:

**Comment by olemis** [comment]
@glyph importing compiled source code works in latest version . However atm we are not distributing a pre-compiled version of stdlib files . I've not tried any tool in particular for doing so either . Nevertheless if that at all matters let's deal with such aspects in a separate ticket . 

BTW there is no guarantee for obtaining performance improvements in execution time after pre-compiling all files in stdlib .

**Comment by glyph** [comment]
I don't see any docs in the PR for how to do the generation of `.pyc.js` files though - did that arrive somewhere else?  Since Brython can't write back to the server, I'm assuming it can't be quite as seamless as Python normally is :).

**Comment by olemis** [comment]
This is the kind of things I wanted to track progress of in a separate ticket ... but I'll reopen this one instead i.e. work in progress .

**Comment by glyph** [comment]
yeah it seems like this should stay open until we have some confirmation that it helps the performance issue.  Still, very excited to see the ongoing work!  thank you!

**Comment by olemis** [comment]
@glyph for converting python code to its javascript equivalent please use `www/tests/precompile.html` tool . It needs further improvements (see #239) , but I consider it a good start for the moment .
